source/Octopus.Tentacle.Tests.Integration/WorkspaceCleanerTests.cs

```diff
@@ -0,0 +1,183 @@
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using FluentAssertions;
+using NUnit.Framework;
+using Octopus.Tentacle.CommonTestUtils.Builders;
+using Octopus.Tentacle.Maintenance;
+using Octopus.Tentacle.Scripts;
+using Octopus.Tentacle.Tests.Integration.Support;
+using Octopus.Tentacle.Tests.Integration.Util;
+using Octopus.Tentacle.Tests.Integration.Util.Builders;
+using Octopus.Tentacle.Tests.Integration.Util.Builders.Decorators;
+
+namespace Octopus.Tentacle.Tests.Integration
+{
+    [IntegrationTestTimeout]
+    public class WorkspaceCleanerTests : IntegrationTest
+    {
+        [Test]
+        [TentacleConfigurations(testDefaultTentacleRuntimeOnly: true)]
+        public async Task WhenScriptServiceIsRunning_ThenWorkspaceIsNotDeleted(TentacleConfigurationTestCase tentacleConfigurationTestCase)
+        {
+            var cleanerDelay = TimeSpan.FromMilliseconds(500);
+            var deleteWorkspacesOlderThan = TimeSpan.FromMilliseconds(500);
+
+            var startScriptCommand = new StartScriptCommandV2Builder().WithScriptBody(b => b.Print("Hello")).Build();
+
+            var workspaceDirectory = string.Empty;
+            await using var clientAndTentacle = await tentacleConfigurationTestCase.CreateBuilder()
+                .WithTentacle(b =>
+                {
+                    b.WithRunTentacleEnvironmentVariable(WorkspaceCleanerConfiguration.CleanerDelayEnvironmentVariableName, cleanerDelay)
+                        .WithRunTentacleEnvironmentVariable(WorkspaceCleanerConfiguration.DeleteWorkspacesOlderThanTimeSpanEnvironmentVariableName, deleteWorkspacesOlderThan);
+                })
+                .WithTentacleServiceDecorator(new TentacleServiceDecoratorBuilder()
+                    .DecorateScriptServiceV2With(d =>
+                        d.BeforeCompleteScript(async (_, _) =>
+                            {
+                                Directory.Exists(workspaceDirectory).Should().BeTrue("Workspace should exist before we complete");
+
+                                await Task.Delay(2 * 500, CancellationToken);
+
+                                Directory.Exists(workspaceDirectory).Should().BeTrue("Workspace should not have been cleaned up");
```

Comment:
Is there a way we can guarantee the cleaner has run between 39 and 43? E.g. At 39, create a dummy Work folder and Output.log and at 43 wait until and assert it was cleaned up. That way we guarantee we are testing the cleaner and it's run as expected.

Reply:
🤔  Creating the file would help. But that only guarantees that it has at least started running, and not necessarily finished.

The only way I can think of to guarantee it has indeed run to completion is to maybe query the system log itself of the running tentacle and count the log lines that show it has finished (and ensure the count has gone up).

But it seems that we currently only log Info level or higher (and these are logged as Verbose) 😑

For now, I've done the dummy file approach, we can discuss this in more detail when you get back 🙂

Reply:
Actually, after I fixed the "Is Running" issue, the "dummy folder" solution can't work anymore. 

Mainly because as soon as the script finishes running, but before "complete script" is called, it now meets the criteria of a workspace that has to be cleaned up.

What stops it from being cleaned up normally is that we wait 1 day before cleaning up candidates. But we had purposely decreased that time to 500ms in this test in order to clean up the "dummy folder". I.e. they both use the same mechanism.

We could add yet another threshold for files that have just completed. But that feels excessive.

For now, I've put it back the way it was, and we can discuss further when you're back.